GitCommands/Remote/GitRemoteManager.cs

```diff
@@ -74,27 +74,25 @@ public GitRemoteManager(Func<IGitModule> getModule)
         [SuppressMessage("ReSharper", "RedundantArgumentDefaultValue")]
         public void ConfigureRemotes(string remoteName)
         {
+            var remoteNameToLower = remoteName.ToLower();
+
             var _module = GetModule();
             var localConfig = _module.LocalConfigFile;
 
             foreach (var remoteHead in _module.GetRefs(true, true))
             {
-                foreach (var localHead in _module.GetRefs(true, true))
+                var remoteHeadToLower = remoteHead.Name.ToLower();
+
+                if (remoteHead.IsRemote ||
+                    remoteHead.IsTag ||
+                    !remoteHeadToLower.Contains(remoteNameToLower) ||
+                    !string.IsNullOrEmpty(remoteHead.GetTrackingRemote(localConfig)))
```

Comment:
The new code doesn't seem to be equivalent to the original. In the original implementation the check is `if (!remoteHead.IsRemote...` and in the new one it has been inverted.

Reply:
Hello, if you really check the 3 commits that I did it cocluded in this. Could you check my second step commit? I'm aware that the if statements are gone. That's why I did it in three step commit to make it possible to discuss and revert to previous refactor iteration.

Reply:
Would you like to check this diff [Second step of refactoring.](https://github.com/gitextensions/gitextensions/pull/4531/commits/7b27ba13dc09fe26376c1d4d3664b8334b9f75e5) ?

Thanks for your time.

Reply:
Your change looks as though it may logically equal but I am not 100% certain - there are too many moving parts. Currently cyclomatic complexity is very high:
![image](https://user-images.githubusercontent.com/4403806/36670209-a37fdfc0-1b4b-11e8-9819-0dbcddb2db10.png)
Without tests, which can guarantee parity, I can't accept the change.

As you have rightly identified - `GetRefs` call results must be reused. 
You have also detected that further optimisations can be made. 👍 

I don't necessarily agree with caching `ToLower` call results:
* [`ToUpper` is allegedly faster](https://stackoverflow.com/questions/9033/hidden-features-of-c#12137)
* personally I would use `string.Compare` with `OrdinalIgnoreCase` instead (as [recommended by MS](https://msdn.microsoft.com/en-us/library/ms973919.aspx?f=255&MSPPError=-2147217396)).

It is non critical though, just sharing my thoughts.

Reply:
I didnt use string.Compare because the app uses .Contains and not .Equal.
For the Unit Tests i would like if you can to provide me with cases that i can write as unit tests.
I want that so i can make sure before updating this pull request and waste more of your time.

Thanks

Reply:
Apologies, I meant `IndexOf`. It's past midnight already...

Reply:
@RussKie no worries. I understand, different timezones..

Reply:
> waste more of your time.

If we're moving in the right direction and reducing the chaos, if you're learning and having fun - then in my books it is time well spent.

> can to provide me with cases that i can write as unit tests.

I would like to see a test for each code path, e.g.:
* Should_not_update_localHead_if_remoteHead_is_local
* Should_not_update_localHead_if_localHead_is_remote and so on

Have a go, let me know if you are having troubles and I'll try to help you

Reply:
@RussKie Thanks for your kind words. Basically i'm doing the contributions to learn more about code unit testing and how to handle big projects. So constructive critique about my code and commits is always welcome.

Thank you again.